stubs/docker/docker/types/services.pyi

```diff
@@ -1,170 +1,185 @@
-        tmpfs_size: Incomplete | None = None,
-        tmpfs_mode: Incomplete | None = None,
+        labels: dict[Incomplete, Incomplete] | None = None,
+        driver_config: DriverConfig | None = None,
+        tmpfs_size: int | str | None = None,
+        tmpfs_mode: int | None = None,
     ) -> None: ...
     @classmethod
-    def parse_mount_string(cls, string): ...
+    def parse_mount_string(cls, string: str) -> Mount: ...
 
 class Resources(dict[str, Incomplete]):
     def __init__(
         self,
-        cpu_limit: Incomplete | None = None,
-        mem_limit: Incomplete | None = None,
-        cpu_reservation: Incomplete | None = None,
-        mem_reservation: Incomplete | None = None,
-        generic_resources: Incomplete | None = None,
+        cpu_limit: int | None = None,
+        mem_limit: int | None = None,
+        cpu_reservation: int | None = None,
+        mem_reservation: int | None = None,
+        generic_resources: dict[str, Incomplete] | list[str] | None = None,
     ) -> None: ...
 
 class UpdateConfig(dict[str, Incomplete]):
     def __init__(
         self,
         parallelism: int = 0,
-        delay: Incomplete | None = None,
-        failure_action: str = "continue",
-        monitor: Incomplete | None = None,
-        max_failure_ratio: Incomplete | None = None,
-        order: Incomplete | None = None,
+        delay: int | None = None,
+        failure_action: Literal["pause", "continue", "rollback"] = "continue",
+        monitor: int | None = None,
+        max_failure_ratio: float | None = None,
+        order: Literal["start-first", "stop-first"] | None = None,
     ) -> None: ...
 
 class RollbackConfig(UpdateConfig): ...
 
 class RestartConditionTypesEnum:
-    NONE: Incomplete
-    ON_FAILURE: Incomplete
-    ANY: Incomplete
+    NONE: str
+    ON_FAILURE: str
+    ANY: str
 
 class RestartPolicy(dict[str, Incomplete]):
     condition_types: type[RestartConditionTypesEnum]
-    def __init__(self, condition="none", delay: int = 0, max_attempts: int = 0, window: int = 0) -> None: ...
+    def __init__(
+        self, condition: Literal["none", "on-failure", "any"] = "none", delay: int = 0, max_attempts: int = 0, window: int = 0
+    ) -> None: ...
 
 class DriverConfig(dict[str, Incomplete]):
-    def __init__(self, name, options: Incomplete | None = None) -> None: ...
+    def __init__(self, name: str, options: dict[Incomplete, Incomplete] | None = None) -> None: ...
 
 class EndpointSpec(dict[str, Incomplete]):
-    def __init__(self, mode: Incomplete | None = None, ports: Incomplete | None = None) -> None: ...
+    def __init__(self, mode: str | None = None, ports: dict[str, str | tuple[str | None, ...]] | None = None) -> None: ...
 
-def convert_service_ports(ports): ...
+@overload
+def convert_service_ports(ports: list[_T]) -> list[_T]: ...
+@overload
+def convert_service_ports(ports: dict[str, str | tuple[str | None, ...]]) -> list[dict[str, str | Incomplete]]: ...
```

Comment:
Pretty sure the value is always a string looking at the implementation
```suggestion
def convert_service_ports(ports: dict[str, str | tuple[str | None, ...]]) -> list[dict[str, str]]: ...
```

Reply:
I think so too but decided not to limit it if I'm not sure about it, but if you're sure then I'll fix it